File: swim-java/swim-core-java/swim.api/src/main/java/swim/api/downlink/ListDownlinkRecord.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.api.downlink;

import java.util.Iterator;
import swim.structure.Item;
import swim.structure.Text;
import swim.structure.Value;
import swim.util.Cursor;

public class ListDownlinkRecord extends DownlinkRecord {
  protected final ListDownlink<Value> downlink;

  public ListDownlinkRecord(ListDownlink<Value> downlink) {
    this.downlink = downlink;
  }

  @Override
  public ListDownlink<Value> downlink() {
    return this.downlink;
  }

  @Override
  public boolean isEmpty() {
    return this.downlink.isEmpty();
  }

  @Override
  public boolean isArray() {
    return true;
  }

  @Override
  public boolean isObject() {
    return this.downlink.isEmpty();
  }

  @Override
  public int size() {
    return this.downlink.size();
  }

  @Override
  public boolean containsKey(Value key) {
    return false;
  }

  @Override
  public boolean containsKey(String key) {
    return false;
  }

  @Override
  public Value get(Value key) {
    return Value.absent();
  }

  @Override
  public Value get(String key) {
    return Value.absent();
  }

  @Override
  public Value getAttr(Text key) {
    return Value.absent();
  }

  @Override
  public Value getAttr(String key) {
    return Value.absent();
  }

  @Override
  public Value getSlot(Value key) {
    return Value.absent();
  }

  @Override
  public Value getSlot(String key) {
    return Value.absent();
  }

  @Override
  public Item get(int index) {
    if (0 <= index && index < this.downlink.size()) {
      return this.downlink.get(index);
    }
    return Item.absent();
  }

  @Override
  public Item getItem(int index) {
    return this.downlink.get(index);
  }

  @Override
  public Value put(Value key, Value newValue) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Value put(String key, Value newValue) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Value putAttr(Text key, Value newValue) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Value putAttr(String key, Value newValue) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Value putSlot(Value key, Value newValue) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Value putSlot(String key, Value newValue) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Item setItem(int index, Item newItem) {
    return this.downlink.set(index, newItem.toValue());
  }

  @Override
  public boolean add(Item item) {
    return this.downlink.add(item.toValue());
  }

  @Override
  public void add(int index, Item item) {
    this.downlink.add(index, item.toValue());
  }

  @Override
  public Item remove(int index) {
    return this.downlink.remove(index);
  }

  @Override
  public boolean removeKey(Value key) {
    return false;
  }

  @Override
  public boolean removeKey(String key) {
    return false;
  }

  @Override
  public void clear() {
    this.downlink.clear();
  }

  @Override
  public Iterator<Value> keyIterator() {
    return Cursor.empty();
  }
}

File: swim-js/swim-runtime-js/swim-host-js/@swim/client/main/downlink/ListDownlinkRecord.ts
// Copyright 2015-2021 Swim Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Cursor} from "@swim/util";
import {AnyItem, Item, AnyValue, Value, Record, AnyText, AnyNum, Num} from "@swim/structure";
import {DownlinkRecord} from "./DownlinkRecord";
import type {ListDownlink} from "./ListDownlink";

export class ListDownlinkRecord extends DownlinkRecord {
  constructor(downlink: ListDownlink<Value, AnyValue>) {
    super();
    Object.defineProperty(this, "downlink", {
      value: downlink,
      enumerable: true,
    });
  }

  override readonly downlink!: ListDownlink<Value, AnyValue>;

  override isEmpty(): boolean {
    return this.downlink.isEmpty();
  }

  override isArray(): boolean {
    return true;
  }

  override isObject(): boolean {
    return this.downlink.isEmpty();
  }

  override get length(): number {
    return this.downlink.length;
  }

  override has(key: AnyValue): boolean {
    return false;
  }

  override get(): Record;
  override get(key: AnyValue): Value;
  override get(key?: AnyValue): Value {
    if (key === void 0) {
      return this;
    } else {
      return Value.absent();
    }
  }

  override getAttr(key: AnyText): Value {
    return Value.absent();
  }

  override getSlot(key: AnyValue): Value {
    return Value.absent();
  }

  override getItem(index: AnyNum): Item {
    if (index instanceof Num) {
      index = index.value;
    }
    const n = this.downlink.length;
    if (index < 0) {
      index = n + index;
    }
    index = Math.min(Math.max(0, index), n - 1);
    if (index >= 0) {
      return this.downlink.get(index);
    }
    return Item.absent();
  }

  override set(key: AnyValue, newValue: AnyValue): this {
    throw new Error("unsupported");
  }

  override setAttr(key: AnyText, newValue: AnyValue): this {
    throw new Error("unsupported");
  }

  override setSlot(key: AnyValue, newValue: AnyValue): this {
    throw new Error("unsupported");
  }

  override setItem(index: AnyNum, newItem: AnyItem): this {
    if (index instanceof Num) {
      index = index.value;
    }
    const n = this.downlink.length;
    if (index < 0) {
      index = n + index;
    }
    index = Math.min(Math.max(0, index), n - 1);
    if (index >= 0) {
      this.downlink.set(index, Value.fromAny(newItem));
    }
    return this;
  }

  override push(...newItems: AnyItem[]): number {
    return this.downlink.push(...newItems);
  }

  override splice(start: number, deleteCount?: number, ...newItems: AnyItem[]): Item[] {
    return this.downlink.splice(start, deleteCount, ...newItems);
  }

  override delete(key: AnyValue): Item {
    return Item.absent();
  }

  override clear(): void {
    this.downlink.clear();
  }

  override forEach<T>(callback: (item: Item, index: number) => T | void): T | undefined;
  override forEach<T, S>(callback: (this: S, item: Item, index: number) => T | void,
                         thisArg: S): T | undefined;
  override forEach<T, S>(callback: (this: S | undefined, item: Item, index: number) => T | void,
                         thisArg?: S): T | undefined {
    return this.downlink.forEach(callback, thisArg);
  }

  override keyIterator(): Cursor<Value> {
    return Cursor.empty();
  }
}
